crates/ruff_linter/src/rules/ruff/rules/sort_dunder_slots.rs

```diff
@@ -0,0 +1,340 @@
+                    } else {
+                        return None;
+                    }
+                }
+                // If `None` was present in the keys, it indicates a "** splat", .e.g
+                // `__slots__ = {"foo": "bar", **other_dict}`
+                // If `None` wasn't present in the keys,
+                // the length of the keys should always equal the length of the values
+                assert_eq!(narrowed_keys.len(), values.len());
+                let display_kind = DisplayKind::Dict { values };
+                Self {
+                    elts: Cow::Owned(narrowed_keys),
+                    range: *range,
+                    display_kind,
+                }
+            }
+            _ => return None,
+        };
+        Some(result)
+    }
+
+    fn generate_fix(&self, items: &[&str], checker: &Checker) -> Option<Fix> {
+        let locator = checker.locator();
+        let is_multiline = locator.contains_line_break(self.range());
+        let sorted_source_code = {
+            if is_multiline {
+                // Sorting multiline dicts is unsupported
+                let display_kind = self.display_kind.as_sequence()?;
+                let analyzed_sequence = MultilineStringSequenceValue::from_source_range(
+                    self.range(),
+                    display_kind,
+                    locator,
+                )?;
+                assert_eq!(analyzed_sequence.len(), self.elts.len());
+                analyzed_sequence.into_sorted_source_code(SORTING_STYLE, locator, checker.stylist())
+            } else {
+                match &self.display_kind {
+                    DisplayKind::Dict { values } => {
+                        sort_single_line_elements_dict(&self.elts, items, values, locator)
+                    }
+                    DisplayKind::Sequence(sequence_kind) => sort_single_line_elements_sequence(
+                        sequence_kind,
+                        &self.elts,
+                        items,
+                        locator,
+                        SORTING_STYLE,
+                    ),
+                }
+            }
+        };
+        Some(Fix::safe_edit(Edit::range_replacement(
+            sorted_source_code,
+            self.range,
+        )))
+    }
+}
+
+struct DictElements<'a>(Vec<(&'a &'a str, &'a ast::Expr, &'a ast::Expr)>);
+
+impl<'a> DictElements<'a> {
+    fn new(elements: &'a [&str], key_elts: &'a [ast::Expr], value_elts: &'a [ast::Expr]) -> Self {
+        assert_eq!(key_elts.len(), elements.len());
+        assert_eq!(elements.len(), value_elts.len());
+        assert!(
+            elements.len() >= 2,
+            "A sequence with < 2 elements cannot be unsorted"
+        );
+        Self(izip!(elements, key_elts, value_elts).collect())
+    }
+
+    fn last_item_index(&self) -> usize {
+        // Safe from underflow, as the constructor guarantees
+        // that the underlying vector has length >= 2
+        self.0.len() - 1
+    }
```

Comment:
I feel like I'd find it clearer to just have `.len()` on this and do `if i < element_trios.len() - 1`, since that's such a common pattern in iteration.

Reply:
I definitely know what you mean, but here specifically I do quite like having this logic right next to the assertions in the constructor method that maintain the invariant that protect us from underflow here